npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -57,14 +60,135 @@ var (
+	- if flush fails because the set doesn't exist (should never happen because we're listing sets right before), then ignore it and the destroy
+	- if flush fails otherwise, then add to destroyFailureCount and continue (aborting the destroy too)
+	- if destroy fails because the set doesn't exist (should never happen since the flush operation would have worked), then ignore it
+	- if destroy fails for another reason, then ignore it and add to destroyFailureCount and mark for reconcile (TODO)
+
+	example:
+		grep output:
+			azure-npm-123456
+			azure-npm-987654
+			azure-npm-777777
+
+		example restore file [flag meanings: -F (flush), -X (destroy)]:
+			-F azure-npm-123456
+			-F azure-npm-987654
+			-F azure-npm-777777
+			-X azure-npm-123456
+			-X azure-npm-987654
+			-X azure-npm-777777
+
+	prometheus metrics:
+		After this function, NumIPSets should be 0 or the number of NPM IPSets that existed and failed to be destroyed.
+		When NPM restarts, Prometheus metrics will initialize at 0, but NPM IPSets may exist.
+		We will reset ipset entry metrics if the restore succeeds whether or not some flushes/destroys failed (NOTE: this is different behavior than v1).
+		If a flush fails, we could update the num entries for that set, but that would be a lot of overhead.
+*/
 func (iMgr *IPSetManager) resetIPSets() error {
-	// called on failure or when NPM is created
-	// so no ipset cache. need to use ipset list like in ipsm.go
+	listCommand := iMgr.ioShim.Exec.Command(ipsetCommand, ipsetListFlag, ipsetNameFlag)
+	grepCommand := iMgr.ioShim.Exec.Command(ioutil.Grep, azureNPMPrefix)
+	azureIPSets, haveAzureIPSets, commandError := ioutil.PipeCommandToGrep(listCommand, grepCommand)
+	if commandError != nil {
+		return npmerrors.SimpleErrorWrapper("failed to run ipset list for resetting IPSets", commandError)
+	}
+	if !haveAzureIPSets {
+		metrics.ResetNumIPSets()
+		metrics.ResetIPSetEntries()
+		return nil
+	}
+	creator, originalNumAzureSets, destroyFailureCount := iMgr.fileCreatorForReset(azureIPSets)
+	restoreError := creator.RunCommandWithFile(ipsetCommand, ipsetRestoreFlag)
+	if restoreError != nil {
+		metrics.SetNumIPSets(originalNumAzureSets)
+		// NOTE: the num entries for sets may be incorrect if the restore fails
+		return npmerrors.SimpleErrorWrapper("failed to run ipset restore for resetting IPSets", restoreError)
+	}
+	if metrics.NumIPSetsIsPositive() {
+		metrics.SetNumIPSets(*destroyFailureCount)
+	} else {
+		metrics.ResetNumIPSets()
+	}
+	metrics.ResetIPSetEntries() // NOTE: the num entries for sets that fail to flush may be incorrect after this
+	return nil
+}
 
-	// create restore file that flushes all sets, then deletes all sets
-	// technically don't need to flush a hashset
+// this needs to be a separate function because we need to check creator contents in UTs
+func (iMgr *IPSetManager) fileCreatorForReset(ipsetListOutput []byte) (creator *ioutil.FileCreator, originalNumAzureSets int, destroyFailureCount *int) {
```

Comment:
as discussed offline avoid usage of named returns, and potentially just return creator, originalNumAzureSets and destroyFailureCounts are purely used for for metrics tracking and don't particularly related to the usage of creator in clients

Reply:
should we fail the lint check in favor of unnamed returns? Also, I'm not sure how we could decouple the numbers for metrics from the function output because we set metrics based on the success/failure of `creator.RunCommand`